File: App/backend/schemas/presets.py
"""Pydantic schemas for prompt preset management."""
from datetime import datetime
from typing import List, Optional, Union, Dict, Any, Literal

from pydantic import BaseModel, Field, ConfigDict

from .sub_agents import SubAgentDefinition


class PresetCreate(BaseModel):
    """Create new preset"""
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    initialize_with_defaults: bool = Field(True, description="Create with default prompts and fragments")


class PresetUpdate(BaseModel):
    """Update preset metadata"""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)


class PresetDuplicateRequest(BaseModel):
    """Request to duplicate a preset"""
    new_name: str = Field(..., min_length=1, max_length=100)
    new_description: Optional[str] = Field(None, max_length=500)


class PresetListItem(BaseModel):
    """Preset item for list responses"""
    id: str
    name: str
    description: Optional[str]
    is_active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class PresetResponse(BaseModel):
    """Full preset details"""
    id: str
    name: str
    description: Optional[str]
    prompt_count: int
    fragment_count: int
    variable_count: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class PresetDetailResponse(BaseModel):
    """Preset with full details including active status"""
    id: str
    name: str
    description: Optional[str]
    is_active: bool
    prompt_count: int
    fragment_count: int
    variable_count: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ActivePresetResponse(BaseModel):
    """Response for active preset info"""
    id: str
    name: str
    description: Optional[str]

    class Config:
        from_attributes = True


# --- Export/Import Schemas ---

class ExportPromptData(BaseModel):
    """Single prompt content for export"""
    model_config = ConfigDict(extra="forbid")
    content: str


class ExportFragmentData(BaseModel):
    """Single fragment data for export"""
    model_config = ConfigDict(extra="forbid")
    content: str
    description: Optional[str] = None


class ExportVariableData(BaseModel):
    """Single variable data for export"""
    model_config = ConfigDict(extra="forbid")
    name: str
    var_type: str
    value: Union[str, int, float, bool, None]
    select_options: Optional[List[str]] = None
    number_options: Optional[Dict[str, Any]] = None
    description: Optional[str] = None
    display_order: int


class PresetExportData(BaseModel):
    """Complete preset export structure"""
    model_config = ConfigDict(extra="forbid")
    format_version: Literal[1]
    preset: Dict[str, Optional[str]]
    prompts: Dict[str, Any]
    fragments: Dict[str, Dict[str, ExportFragmentData]]
    variables: List[ExportVariableData]
    sub_agents: List[SubAgentDefinition]
    exported_at: str


class PresetImportData(BaseModel):
    """Request to import a preset from exported JSON"""
    model_config = ConfigDict(extra="forbid")
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    data: PresetExportData

File: App/frontend/src/types/presets.ts
/**
 * Types for prompt preset management
 */

import type { SubAgentDefinition } from './subAgents';
import type { ScenarioDocument } from './scenarios';

/**
 * Preset list item (minimal info for list display)
 */
export interface PresetListItem {
  id: string;
  name: string;
  description: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

/**
 * Full preset response with counts
 */
export interface PresetResponse {
  id: string;
  name: string;
  description: string | null;
  prompt_count: number;
  fragment_count: number;
  variable_count: number;
  created_at: string;
  updated_at: string;
}

/**
 * Preset with full details including active status
 */
export interface PresetDetailResponse {
  id: string;
  name: string;
  description: string | null;
  is_active: boolean;
  prompt_count: number;
  fragment_count: number;
  variable_count: number;
  created_at: string;
  updated_at: string;
}

/**
 * Active preset info response
 */
export interface ActivePresetResponse {
  id: string;
  name: string;
  description: string | null;
}

/**
 * Request to create a new preset
 */
export interface PresetCreate {
  name: string;
  description?: string;
  initialize_with_defaults?: boolean;
}

/**
 * Request to update preset metadata
 */
export interface PresetUpdate {
  name?: string;
  description?: string;
}

/**
 * Request to duplicate a preset
 */
export interface PresetDuplicateRequest {
  new_name: string;
  new_description?: string;
}

// --- Export/Import Types ---

/**
 * Exported fragment data structure
 */
export interface ExportFragmentData {
  content: string;
  description?: string | null;
}

/**
 * Exported variable data structure
 */
export interface ExportVariableData {
  name: string;
  var_type: 'string' | 'number' | 'boolean' | 'select';
  value: string | number | boolean | null;
  select_options?: string[] | null;
  number_options?: {
    min?: number | null;
    max?: number | null;
    step?: number | null;
    input_type?: 'input' | 'slider';
  } | null;
  description?: string | null;
  display_order: number;
}

/**
 * Complete preset export structure
 */
export interface PresetExportData {
  format_version: 1;
  preset: {
    name: string;
    description: string | null;
  };
  // Scenario v1: prompts payload is scenario documents per (task_type, task_subtype)
  prompts: Record<string, Record<string, ScenarioDocument>>;
  fragments: Record<string, Record<string, ExportFragmentData>>;
  variables: ExportVariableData[];
  sub_agents: SubAgentDefinition[];
  exported_at: string;
}

/**
 * Request to import a preset
 */
export interface PresetImportRequest {
  name: string;
  description?: string;
  data: PresetExportData;
}
